### PR TITLE
Execute service error handler on panic

### LINKED ISSUE #34: Service error handler not run in multiplexing client when the consumer codec panics
Using tokio-tower's multiplexing [client](https://github.com/tower-rs/tokio-tower/blob/master/src/multiplex/client.rs#L151), if the consumer codec panics, the service error handler is not run. This might corrupt some state on the application side that might rely on a connection close to trigger the service error handler, for example to remove a connection from their own pool.

What do you think about using RAII in the [transport task](https://github.com/tower-rs/tokio-tower/blob/master/src/multiplex/client.rs#L351) to execute the service error handler on panic?

### PR DESCRIPTION
**Pull request by estaban**
Fixes #34

### CODE REVIEW

File: src/multiplex/client.rs [hunk 1]
```diff
@@ -323,6 +323,30 @@ where
     }
 }
 
+/// Handles executing the service error handler in case awaiting the `ClientInner` Future panics.
+struct ClientInnerCleanup<Request, T, E, F>
+where
+    T: Sink<Request> + TryStream,
+    E: From<Error<T, Request>>,
+    F: FnOnce(E),
```
Comment (LucioFranco) [inline src/multiplex/client.rs:331]: why are all these bounds needed here? I can see why they might be needed in the Drop impl.

File: src/multiplex/client.rs [hunk 2]
```diff
@@ -323,6 +323,30 @@ where
...
+
+impl<Request, T, E, F> Drop for ClientInnerCleanup<Request, T, E, F>
+where
+    T: Sink<Request> + TryStream,
+    E: From<Error<T, Request>>,
+    F: FnOnce(E),
+{
+    fn drop(&mut self) {
+        if let Some(handler) = self.on_service_error.take() {
+            (handler)(E::from(Error::<T, Request>::TransportUnexpectedExit))
```
Comment (LucioFranco) [inline src/multiplex/client.rs:345]: I wonder if we need to address the case where we are already panicking and the closure panics. Though I can't find the reference code I remember writing for this style of thing. So I think its fine but may be worth some sort of doc comment.

Panic inside a panic causes an abort.

File: src/error.rs [hunk 3]
```diff
@@ -29,6 +29,10 @@ where
 
     /// The server sent a response that the client was not expecting.
     Desynchronized,
+
+    /// The underlying transport task did not exit gracefully (either panic or cancellation).
+    /// Transport task panics can happen for example when the codec logic panics.
+    TransportUnexpectedExit,
```
Comment (LucioFranco) [inline src/error.rs:35]: Nit: maybe call this `TransportDropped` since that tracks closer to what the actually issue is?

**Review by LucioFranco — COMMENTED** [review]
LGTM left two comments but I think we can merge this concept.

File: src/multiplex/client.rs (hunk 2 above)
Comment (estaban) [reply src/multiplex/client.rs]: Added documentation about the fact that panicking within the closure will abort

File: src/error.rs (hunk 3 above)
Comment (estaban) [reply src/error.rs]: Sounds good, done

File: src/multiplex/client.rs (hunk 1 above)
Comment (LucioFranco) [reply src/multiplex/client.rs:334]: Oh yeah you're right...I am not sure why but I assume there is probably a good reason.

**Review by LucioFranco — APPROVED** [review]

**Pull request merged**